src/jit/ssabuilder.cpp

```diff
@@ -673,6 +673,99 @@ static GenTree* GetPhiNode(BasicBlock* block, unsigned lclNum)
     return nullptr;
 }
 
+//------------------------------------------------------------------------
+// InsertPhi: Insert a new GT_PHI statement.
+//
+// Arguments:
+//    block  - The block where to insert the statement
+//    lclNum - The variable number
+//
+void SsaBuilder::InsertPhi(BasicBlock* block, unsigned lclNum)
+{
+    var_types type = m_pCompiler->lvaGetDesc(lclNum)->TypeGet();
+
+    GenTree* lhs = m_pCompiler->gtNewLclvNode(lclNum, type);
+    // PHIs and all the associated nodes do not generate any code so the costs are always 0
+    lhs->SetCosts(0, 0);
+    GenTree* phi = new (m_pCompiler, GT_PHI) GenTreePhi(type);
+    phi->SetCosts(0, 0);
+    GenTree* asg = m_pCompiler->gtNewAssignNode(lhs, phi);
+    // Evaluate the assignment RHS (the PHI node) first. This way the LHS will end up right
+    // in front of the assignment in the linear order, that ensures that using gtGetParent
+    // on the LHS to find the assignment doesn't have to traverse the PHI and its args.
+    asg->gtFlags |= GTF_REVERSE_OPS;
+    asg->SetCosts(0, 0);
+
+    // Create the statement and chain everything in linear order - PHI, LCL_VAR, ASG
+    GenTreeStmt* stmt = m_pCompiler->gtNewStmt(asg);
+    stmt->gtStmtList  = phi;
+    phi->gtNext       = lhs;
+    lhs->gtPrev       = phi;
+    lhs->gtNext       = asg;
+    asg->gtPrev       = lhs;
+
+#ifdef DEBUG
+    unsigned seqNum = 1;
+    for (GenTree* node = stmt->AsStmt()->gtStmtList; node != nullptr; node = node->gtNext)
+    {
+        node->gtSeqNum = seqNum++;
+    }
+#endif // DEBUG
+
+    m_pCompiler->fgInsertStmtAtBeg(block, stmt);
+
+    JITDUMP("Added PHI definition for V%02u at start of " FMT_BB ".\n", lclNum, block->bbNum);
+}
+
+//------------------------------------------------------------------------
+// AddPhiArg: Add a new GT_PHI_ARG node to an existing GT_PHI node.
+//
+// Arguments:
+//    block  - The block that contains the statement
+//    stmt   - The statement that contains the GT_PHI node
+//    lclNum - The variable number
+//    ssaNum - The SSA number
+//    pred   - The predecessor block
+//
+void SsaBuilder::AddPhiArg(
+    BasicBlock* block, GenTreeStmt* stmt, GenTreePhi* phi, unsigned lclNum, unsigned ssaNum, BasicBlock* pred)
+{
+#ifdef DEBUG
+    // Make sure it isn't already present: we should only add each definition once.
+    for (GenTreePhi::Use& use : phi->Uses())
+    {
+        assert(use.op->AsPhiArg()->GetSsaNum() != ssaNum);
```

Comment:
Perhaps also assert that it's not for the same `pred`?

Reply:
>  Speaking of which, do you have data on how large these lists get in bad cases?

Here you go: https://1drv.ms/x/s!Av4baJYSo5pjgsQZfywTf5l3NUXitw 

Data from all FX assemblies. Largest PHI has 255 operands. 90% have only 2 operands.

I was actually considering handling the 2 operands case specially - store the 2 uses inside the `GenTreePhi` node itself. But the code required to handle this special case gets kind of ugly.

Reply:
> It is more operands in some cases but perhaps avoids quadratic builder behavior for large phis.

I was curious to see what such quadratic behavior means on today hardware: https://gist.github.com/mikedn/23e708655cdb1d431ed165e300192c9e

It's less than 1μs for N = 10, ~68μs for N = 255, ~1ms for N = 1000 and ~200ms for N = 10000.

I'd say we're on the safe side.

Reply:
> And if we just record the "first block" that brings in a value I wonder how RangeCheck::ComputeRange works correctly for PHIs as it seems like it would skip looking at some preds.

Indeed, that seems broken. If one predecessor asserts that `i <= 2` and another asserts that `i <= 3` the resulting PHI range should have upper bound 3. But if you miss the later predecessor then the upper bound will be only 2... Now, there may be something else in RangeCheck that prevents this from happening, though I starred a lot at RangeCheck in the past and I just don't see that something else.

Reply:
> Perhaps also assert that it's not for the same pred?

Nope, it turns out that this assert isn't valid due to exception handling. It is possible to have something like `PHI BB05:V02:u2, BB05:V02:u3, BB05:V02:u4` when the same block in a try region contains more than one definition of the variable that's live in the handler.

It's the BB-SSA_NUM pair that's supposed to be unique in the PHI arg list. But since the JIT already requires SSA_NUM to be unique there's not much we can validate about the BB.